File: frontend/src/interfaces/review.ts
export interface ReviewAuthor {
  id: number;
  first_name: string;
}

export interface ReviewRestaurantBrief {
  id: number;
  name: string;
  slug: string;
  cuisine: string;
  city: string;
  price_range: number;
  rating: number | null;
}

export interface Review {
  id: number;
  user_id: number;
  restaurant_id: number;
  rating: number;
  text: string | null;
  author: ReviewAuthor;
  restaurant: ReviewRestaurantBrief;
  created_at: string;
  updated_at: string | null;
}

export interface ReviewFormData {
  rating: number;
  text: string;
}

File: backend/app/models/review.py
from datetime import datetime

from sqlalchemy import Integer, Text, ForeignKey, DateTime, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.db.database import Base, get_utc_now
from typing import Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from app.models.user import User
    from app.models.restaurant import Restaurant


class Review(Base):
    __tablename__ = "reviews"
    __table_args__ = (
        UniqueConstraint("user_id", "restaurant_id", name="uq_user_restaurant_review"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)

    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id"), nullable=False, index=True
    )
    restaurant_id: Mapped[int] = mapped_column(
        ForeignKey("restaurants.id"), nullable=False, index=True
    )

    rating: Mapped[int] = mapped_column(Integer, nullable=False)
    text: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=get_utc_now, nullable=False
    )
    updated_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime, onupdate=get_utc_now, nullable=True
    )

    user: Mapped["User"] = relationship("User", back_populates="reviews")
    restaurant: Mapped["Restaurant"] = relationship(
        "Restaurant", back_populates="reviews"
    )
